src/catchup/VerifyLedgerChainWork.cpp

```diff
@@ -84,10 +84,31 @@ verifyLastLedgerInCheckpoint(LedgerHeaderHistoryEntry const& ledger,
     return HistoryManager::VERIFY_STATUS_OK;
 }
 
+template <typename T>
+void
+trySetFuture(std::promise<T>& promise, T value)
```

Comment:
> Instead of using shared_future with promises, I think it might be better to just use std::shared_ptr<std::atomic<bool>>. You never actually access the shared_future in an async way, all accesses are concurrent, so you're not benefiting from the async interface. std::atomic<bool> can be set multiple times and we wouldn't have to catch any exceptions. You could just have a single assert in the constructor that the initial shared_ptr passed to the constructor in non-null and then you should be good to go without any further error checking.

So I actually had a similar debate :) I agree that we aren't doing anything async, and perhaps using futures is a bit of an overkill, but I picked it for two reasons (lmk if you disagree):
* The Work subsystem doesn't really have a good way to share state between works. https://github.com/stellar/stellar-core/pull/2598 introduced the promise/future pattern to partially deal with this problem. So I wanted to stay consistent with the codebase, since in this change we're trying to do the same thing.
* I think futures provide more clarity to the interface: this way we know that it is a responsibility of verify-ledger-chain to properly set the state at the end. Similarly, catchup-work can only check and maybe retrieve the state, but it can't make any modifications to it. 

> Currently, it's actually possible to miss a fatal error if we have bad local state and if the first history archive we download from is corrupt. On our first download, mHasTrustedHash will be false, so we set the shared_future to false. On the next download assuming the next history archive we try is correct, mHasTrustedHash and mLocalFailure will both be true, but we will be unable to set the fatal flag and would erroneously retry downloading history again a 3rd time. This would only occur if a node tries to download history 3+ times, the first history archive is corrupt, and if local state is also corrupt, so definitely a corner case, but atomic<bool> would help us there.

This is a good point! I _think_ the current implementation should still work though: we don't do retries on verify-ledger-chain, so if we fail the first time with the corrupt archive, catchup would fail and get reset (resetting the future as well) before a new catchup is triggered.

Reply:
Yeah, I see where you're coming from. I don't think it's a huge deal so I'm good with leaving it as is, but using an async interface synchronously that we have to set in a try-catch loop...makes me feel a certain type of way lol.